sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -858,6 +859,30 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+  * Push [[Limit]] operators through [[Join]] operators, iff the join type is outer joins.
+  * Adding extra [[Limit]] operators on top of the outer-side child/children.
+  */
+object PushLimitThroughOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
I am not sure what is the best way to prove it. If we can add extra `limit` below `union all`, `left outer` and `right outer`,  can we add extra `limit` below `full outer`? In the traditional RDBMS, ```full outer = union all(left outer, right outer)```. I am not sure if Spark SQL has the same semantics. 

`(A full outer join B) limit 5` 
= `(A left outer join B) limit 5` `union all` `(A right outer join B) limit 5`
= `((A limit 5) left outer join (B limit 5))` `union all` `((A limit 5) right outer join (B limit 5))` 
= `((A limit 5) full outer join (B limit 5))`


However, inner join has a big issue if we are trying to add extra `limit`. I am not sure my answer is clear.